daemon/logger/ring.go

```diff
@@ -0,0 +1,219 @@
+package logger
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/Sirupsen/logrus"
+	"golang.org/x/net/context"
+)
+
+const (
+	defaultRingMaxSize = 1000000 // 1MB
+)
+
+// RingLogger is a ring buffer that implements the Logger interface.
+// This is used when lossy logging is OK.
+type RingLogger struct {
+	buffer *messageRing
+	ctx    context.Context
+	cancel func()
+	l      Logger
+	logCtx Info
+}
+
+type ringWithReader struct {
+	*RingLogger
+}
+
+func (r *ringWithReader) ReadLogs(cfg ReadConfig) *LogWatcher {
+	reader, ok := r.l.(LogReader)
+	if !ok {
+		// something is wrong if we get here
+		panic("expected log reader")
+	}
+	return reader.ReadLogs(cfg)
+}
+
+func newRingLogger(driver Logger, logCtx Info, maxSize int64) *RingLogger {
+	ctx, cancel := context.WithCancel(context.Background())
+	l := &RingLogger{
+		buffer: newRing(maxSize),
+		ctx:    ctx,
+		cancel: cancel,
+		l:      driver,
+		logCtx: logCtx,
+	}
+	l.run()
+	return l
+}
+
+// NewRingLogger creates a new Logger that is implemented as a RingBuffer wrapping
+// the passed in logger.
+func NewRingLogger(driver Logger, logCtx Info, maxSize int64) Logger {
+	if maxSize < 0 {
+		maxSize = defaultRingMaxSize
+	}
+	l := newRingLogger(driver, logCtx, maxSize)
+	if _, ok := driver.(LogReader); ok {
+		return &ringWithReader{l}
+	}
+	return l
+}
+
+// Log queues messages into the ring buffer
+func (r *RingLogger) Log(msg *Message) error {
+	select {
+	case <-r.ctx.Done():
+		return r.ctx.Err()
+	default:
+	}
+	return r.buffer.Enqueue(msg)
+}
+
+// Name returns the name of the underlying logger
+func (r *RingLogger) Name() string {
+	return r.l.Name()
+}
+
+// Close closes the logger
+func (r *RingLogger) Close() error {
+	r.cancel()
+	r.buffer.Close()
+	// empty out the queue
+	for _, msg := range r.buffer.Drain() {
+		if err := r.l.Log(msg); err != nil {
+			logrus.WithField("driver", r.l.Name()).WithField("container", r.logCtx.ContainerID).Errorf("Error writing log message: %v", r.l)
+			break
+		}
+	}
+	return r.l.Close()
+}
+
+// run consumes messages from the ring buffer and forwards them to the underling
+// logger.
+// This is run in a goroutine when the RingLogger is created
+func (r *RingLogger) run() {
+	go func() {
+		for {
+			select {
+			case <-r.ctx.Done():
+				return
+			default:
+				msg, err := r.buffer.Dequeue()
+				if err != nil {
+					// buffer is closed
+					return
+				}
+				if err := r.l.Log(msg); err != nil {
+					logrus.WithField("driver", r.l.Name()).WithField("container", r.logCtx.ContainerID).Errorf("Error writing log message: %v", r.l)
+				}
+			}
+		}
+	}()
```

Comment:
nit: Maybe `go run()` instead of this closure? That would make it clearer that `newRingLogger` starts a goroutine, and simplify this function.